finality-aleph/src/justification.rs

```diff
@@ -1,70 +1,220 @@
+        keybox: KeyBox,
+        last_finalized: NumberFor<B>,
+        stop_h: NumberFor<B>,
+    ) {
+        let num = notification.number;
+        let block_hash = notification.hash;
+        match AlephJustification::verify::<B, _>(
+            notification.justification,
+            notification.hash,
+            &crate::MultiKeychain::new(keybox),
+        ) {
+            Ok(justification) => {
+                if num > last_finalized && num <= stop_h {
+                    debug!(target: "afa", "Finalizing block {:?} {:?}", num, block_hash);
+                    self.last_finalization = Instant::now();
+                    self.finalize(num, block_hash, justification);
+                } else {
+                    debug!(target: "afa", "Not finalizing block {:?}. Last finalized {:?}, stop_h {:?}", num, last_finalized, stop_h);
+                }
+            }
+            Err(err) => {
+                error!(target: "afa", "Error when verifying justification {:?} for block {:?} {:?}", err, num, block_hash);
+            }
+        }
+    }
+
+    pub(crate) fn handle_requests(&mut self, num: NumberFor<B>) {
+        let current_time = Instant::now();
+        if current_time - self.last_finalization > Duration::from_secs(10)
+            && current_time - self.last_request > Duration::from_secs(10)
+        {
+            self.last_request = current_time;
+            debug!(target: "afa", "Trying to request block {:?}", num);
+            if let Ok(Some(header)) = self.client.header(BlockId::Number(num)) {
+                debug!(target: "afa", "We have block {:?} with hash {:?}. Requesting justification.", num, header.hash());
+                self.network
+                    .request_justification(&header.hash(), *header.number());
+            } else {
+                debug!(target: "afa", "Cancelling request, because we don't have block {:?}.", num);
+            }
         }
     }
 
     pub(crate) async fn run(mut self) {
-        while let Some(notification) = self.justification_rx.next().await {
-            self.finalization_proposals_tx
-                .unbounded_send(notification)
-                .expect("Notification should succeed");
+        loop {
+            let last_finalized_number = self.client.info().finalized_number;
+            let current_session = self.current_session_id(last_finalized_number);
+            let stop_h = (current_session + 1) * self.session_period - 1;
+            let keybox = self.session_keybox(current_session);
+            if keybox.is_none() {
+                error!(target: "afa", "Cannot fetch the keybox for session {:?}. Waiting 500ms and will try again ...", current_session);
+                Delay::new(Duration::from_millis(500)).await;
+                continue;
+            }
+            let keybox = keybox.expect("We loop until this is some.");
+            let mut ticker_fut = Delay::new(Duration::from_millis(1000));
```

Comment:
Hmm, we can I guess, but then the version with select requires you to play with Either::Left/Right which I think makes the code obscure. If there is a way to use timeout without this syntax than I would be happy to change this :P

Reply:
Well, you can use `futures::stream::select(self.import_justification_rx, self.authority_justification_rx).next()`. The select on streams expect that the streams have the same type and returns a stream of elements of that type. This however loses the information about the source of a notifification (import/aggregator). We have the following options:

* accept that this bit of information is lost (it is used only in logs),
* add a field in the notification which tells the source of the notification,
* instead of calling select on receivers, call it on `self.import_justification_rx.inspect(...)` so that we log an appropriate message before handling the justification

Reply:
Ok, this is addressed, I think :)